x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java

```diff
@@ -1643,6 +1671,137 @@ else if ((other instanceof LessThan || other instanceof LessThanOrEqual) &&
 
             return false;
         }
+
+        private static boolean notEqualsIsRemovableFromConjunction(NotEquals notEquals, List<Range> ranges, List<BinaryComparison> bcs) {
+            Object neqVal = notEquals.right().fold();
+            Integer comp;
+
+            // check on "condition-overlapping" ranges:
+            // a != 2 AND 3 < a < 5 -> 3 < a < 5; a != 2 AND 0 < a < 1 -> 0 < a < 1 (discard NotEquals)
+            // a != 2 AND 2 <= a < 3 -> 2 < a < 3; a != 3 AND 2 < a <= 3 -> 2 < a < 3 (discard NotEquals, plus update Range)
+            // a != 2 AND 1 < a < 3 -> nop (do nothing)
+            for (int i = 0; i < ranges.size(); i ++) {
+                Range range = ranges.get(i);
+
+                if (notEquals.left().semanticEquals(range.value())) {
+                    comp = range.lower().foldable() ? BinaryComparison.compare(neqVal, range.lower().fold()) : null;
+                    if (comp != null) {
+                        if (comp <= 0) {
+                            if (comp == 0 && range.includeLower()) { // a != 2 AND 2 <= a < ? -> 2 < a < ?
+                                ranges.set(i, new Range(range.source(), range.value(), range.lower(), false, range.upper(),
+                                    range.includeUpper()));
+                            }
+                            // else: !.includeLower() : a != 2 AND 2 < a < 3 -> 2 < a < 3; or:
+                            // else: comp < 0 : a != 2 AND 3 < a < ? ->  3 < a < ?
+
+                            return true;
+                        } else { // comp > 0 : a != 4 AND 2 < a < ? : can only remove NotEquals if outside the range
+                            comp = range.upper().foldable() ? BinaryComparison.compare(neqVal, range.upper().fold()) : null;
```

Comment:
I find a bit confusing and potentially problematic, in the future if this code will change, re-using the `comp` variable in this block. Initially, it has a value assigned to it in on an upper level, it goes through two `if` branches, then in a separate, distinct conditional branch on an inner level gets another value and is treated similarly as before. I am not suggesting to change this, I am just mentioning it as a nitpick.

Reply:
The result of a `comp` assignment is used right away (i.e. next line) and there only. So I considered unnecessary to declare multiple vars. But I can see how this can be regarded with unease. I'd think that if the usage pattern ever changes, a new var is indeed preferable.